### PR TITLE
Fix Issue 17286 - A function for comparing two digests securely

### PR DESCRIPTION
**Pull request by JackStouffer**
Given two strings that represent two crypto digests, using std.algorithm.equal to compare them leaves your web application open to timing attacks because it has a short circuit, i.e. it returns `false` on the first inequality it finds.

The attack comes from allowing attacker to brute force your HMAC key for example. See this article for more information, evidence that the attack exists, and why Java gets it wrong: https://codahale.com/a-lesson-in-timing-attacks/

The solution is to have a string comparison that will always be constant time given two strings of the same length.

Ping @andralex, this is a name addition.

### CODE REVIEW

**Comment by JackStouffer** [comment]
Examples of this function is other languages: 

https://docs.python.org/2/library/hmac.html#hmac.compare_digest
http://www.rubydoc.info/github/rack/rack/master/Rack/Utils.secure_compare

**Comment by braddr** [comment]
Constant time algorithms are an important category, so I'm in favor of this in spirit, but the placement is questionable and I'd rather the name reflect the constant-ness rather than the secure-ness.

**Comment by quickfur** [comment]
I'm tempted to say this should go in `std.algorithm.comparison`, but then most people who need to know about this function may not see it when looking up digest-related Phobos modules.

But in any case, I think `ConstantTimeCompare` or something along those lines might be a better name for this.

File: std/digest/digest.d [hunk 1]
```diff
@@ -1021,3 +1021,53 @@ if (isDigest!T) : Digest
...
+{
+    if (s1.length != s2.length)
+    {
+        return false;
+    }
+
+    int result = 0;
+    foreach (i; 0 .. s1.length)
+    {
+        result |= s1[i] ^ s2[i];
```
Comment (quickfur) [inline std/digest/digest.d]: This requires a sig constraint that `T` (1) supports the `^` operation, (2) that said operation does what we think it does, and (2) the result can be OR'd into an `int`.

Comment (quickfur) [inline std/digest/digest.d]: For (2), I think we'd have to restrict `T` to built-in POD types.

**Comment by JackStouffer** [comment]
Added constraints. I think that the name is alright but I'm open to suggestions. I would like to know Andrei's opinion on the matter though.

I also don't know of any use for this function outside of digests, so that's why it's where it is. Let me know if there are other applications for this algorithm.

File: std/digest/digest.d [hunk 2]
```diff
@@ -1021,3 +1021,53 @@ if (isDigest!T) : Digest
...
+ * Params:
+ *     s1 = A digest representation
+ *     s2 = A digest representation
+ * Returns:
+ *     `true` if both representations are equal, `false` otherwise
+ * See_Also:
+ *     $(LINK2 https://security.stackexchange.com/a/74552, An in depth explanation of
+ *     the attack).
+ */
+bool secureCompare(T)(T[] s1, T[] s2) if (isIntegral!T || isSomeChar!T)
```
Comment (andralex) [inline std/digest/digest.d]: "compare" in D nomenclature is closer to "cmp" i.e. ordering, so the better name here is `secureEqual`

File: std/digest/digest.d [hunk 3]
```diff
@@ -1021,3 +1021,53 @@ if (isDigest!T) : Digest
...
+{
+    if (s1.length != s2.length)
+    {
+        return false;
```
Comment (andralex) [inline std/digest/digest.d]: not covered, also it strikes me as a bit much to fully brace this

File: std/digest/digest.d [hunk 4]
```diff
@@ -1021,3 +1021,53 @@ if (isDigest!T) : Digest
...
+ * over `==` for comparing verification tokens, as it protects from brute force
+ * timing attacks that can reveal the key.
+ *
+ * Params:
+ *     s1 = A digest representation
+ *     s2 = A digest representation
+ * Returns:
+ *     `true` if both representations are equal, `false` otherwise
+ * See_Also:
+ *     $(LINK2 https://security.stackexchange.com/a/74552, An in depth explanation of
```
Comment (andralex) [inline std/digest/digest.d]: A link to a reputable source is necessary here.

File: std/digest/digest.d [hunk 5]
```diff
@@ -1021,3 +1021,53 @@ if (isDigest!T) : Digest
     ubyte[5] buf;
     assert(hash.peek(buf).toHexString() == "39A34F41");
 }
+
+/**
+ * Securely compares two digest representations. This function is $(BIGOH n)
+ * for all strings that are of equal length. Use of this function is recommended
+ * over `==` for comparing verification tokens, as it protects from brute force
+ * timing attacks that can reveal the key.
```
Comment (andralex) [inline std/digest/digest.d]: "timing attack" has a good Wikipedia entry and a bunch of other reference.

The documentation should include a brief explanation. I had no idea what the intent is after reading it.

**Comment by JackStouffer** [comment]
> return lhs[i .. $] == rhs[i .. $] && lhs[0 .. i].retro == rhs[0 .. i].retro;

According to my benchmark, it's slower

```
$ dmd -O -inline -release test.d && ./test
old same	        538 ms and 300 μs
new same	1 sec, 162 ms, 32 μs, and 2 hnsecs
old different	545 ms, 930 μs, and 4 hnsecs
new different	696 ms, 248 μs, and 7 hnsecs
```

code

```d
import std.stdio;
import std.range;
import std.algorithm;
import std.traits;
import std.container;
import std.meta;
import std.conv;
import std.datetime;

enum testCount = 10_000_000;

bool secureCompare1(T)(T[] s1, T[] s2) if (isIntegral!T || isSomeChar!T)
{
    if (s1.length != s2.length)
        return false;

    int result = 0;
    foreach (i; 0 .. s1.length)
        result |= s1[i] ^ s2[i];

    return result == 0;
}

bool secureCompare2(T)(T[] lhs, T[] rhs) if (isIntegral!T || isSomeChar!T)
{
    import std.random : uniform;

    if (lhs.length != rhs.length)
        return false;

    immutable i = uniform(0, lhs.length);
    return lhs[i .. $] == rhs[i .. $] && lhs[0 .. i].retro == rhs[0 .. i].retro;
}

void main()
{
    auto hex1 = "02CA3484C375EDD3C0F08D3F50D119E610779018";
    auto hex2 = "02CA3484C375EDD3C0F08D3F50D119E610779018";
    auto hex3 = "A10BBD48B3BBD5765EF2F7CAED0D6BE40A156AE7";

    auto result = to!Duration(benchmark!(() => secureCompare1(hex1, hex2))(testCount)[0]);
    auto result2 = to!Duration(benchmark!(() => secureCompare2(hex1, hex2))(testCount)[0]);

    writeln("old same", "\t", result);
    writeln("new same", "\t", result2);

    auto result3 = to!Duration(benchmark!(() => secureCompare1(hex1, hex3))(testCount)[0]);
    auto result4 = to!Duration(benchmark!(() => secureCompare2(hex1, hex3))(testCount)[0]);

    writeln("old different", "\t", result3);
    writeln("new different", "\t", result4);
}
```

File: std/digest/digest.d (hunk 5 above)
Comment (JackStouffer) [inline std/digest/digest.d]: Unfortunately, the Wikipedia article does not explain the specific attack in question, so I didn't use it.

I'll add a bit more explaination

Comment (andralex) [inline std/digest/digest.d]: Cool thx. You may want to search for one or more other references, the more authoritative the better.

**Comment by andralex** [comment]
Yah no surprise it's slower for short inputs. I guess I'm fine assuming the prevalent inputs are SHA hashes. Speaking of which, we should make this work with ranges.

**Comment by JackStouffer** [comment]
>we should make this work with ranges

will do

**Comment by JackStouffer** [comment]
Updated the docs and made it work with ranges.

**Comment by quickfur** [comment]
@andralex  Your alternative will not thwart a determined attacker if the (P)RNG you use to choose `i` is not cryptographically secure.   Namely, using anything from `std.random` will not work.

File: std/digest/digest.d [hunk 6]
```diff
@@ -1021,3 +1022,150 @@ if (isDigest!T) : Digest
...
+    }
+    else
+    {
+        // Generic case, walk both ranges
+        for (; !r1.empty; r1.popFront(), r2.popFront())
+        {
+            if (r2.empty) return false;
+            result |= r1.front ^ r2.front;
+        }
+        if (!r2.empty) return false;
```
Comment (andralex) [inline std/digest/digest.d]: There's a simpler way to write this loop:

```d
for (;; r1.popFront(), r2.popFront())
{
    if (r1.empty) return r2.empty;
    if (r2.empty) return false;
    result |= r1.front ^ r2.front;
}
```

Eh, about the same :)

**Comment by JackStouffer** [comment]
Thanks!

**Comment by andralex** [comment]
@quickfur here's the code I used. It's 2x slower on the passing comparison and some 15% faster on the failing comparison:

```d
bool secureCompare2(T)(T[] lhs, T[] rhs) if (isIntegral!T || isSomeChar!T)
{
    import std.random;
    static Xorshift32 rnd;
    static bool init;
    if (!init)
    {
        rnd = typeof(rnd)(unpredictableSeed);
        init = true;
    }

    if (lhs.length != rhs.length)
        return false;

    immutable i = uniform(0, lhs.length, rnd);
    return lhs[i .. $] == rhs[i .. $] && lhs[0 .. i].retro == rhs[0 .. i].retro;
}
```

Now, assume `unpredictableSeed` is cryptographically secure. Would that make this algorithm cryptographically secure?

**Comment by jpf91** [comment]
> Now, assume unpredictableSeed is cryptographically secure. Would that make this algorithm cryptographically secure?

I'd say better safe than sorry. Side channel attacks are a difficult problem and unless we have some real crypto experts verify our custom algorithms I would always prefer the common & safe implementation of constant time algorithms. In crypto simple to verify & review is just as important as performance.

BTW @JackStouffer do you have any expert knowledge of crypto stuff? I studied some cryptography textbooks lately and some real world cryptography libraries and I think the API in `std.digest` is not suitable for cryptography related purposes by design. The main problem is that we return the digests by value so we may create copies on the stack in all kinds of places and we can't properly zero/clean all of these copies. This is probably the reason why libraries like `libsodium` expect the caller to allocate storage for the digest and then always use pass by reference semantics. (Sometimes digest outputs are used to derive private keys/ as private keys and therefore having these somewhere in old stack memory is a security issue). What do you think @JackStouffer ?

We also have similar issues for `std.stdio.File / readFile` when reading private key files. I wouldn't trust anything range based or `std.stdio` based not to leak parts of the key file into some internal buffer...

**Comment by JackStouffer** [comment]
I'm by no means an expert. I came across information on this issue when implementing HMAC signing of messages in my project and was surprised that D didn't have a solution.

Could you describe or link to some info on the copying issue? Won't this also be an issue in any scripting language?

**Comment by jpf91** [comment]
> Could you describe or link to some info on the copying issue?

I don't know any link directly covering this so I'll try to explain:

First we have to assume that we want to keep the hash secret. Private or symmetric encryption keys are often derived from hashes (HKDF, used in the noise protocol, ...).

Let's see how libsodium handles this: It has extra functions for [secure memory](https://download.libsodium.org/doc/helpers/memory_management.html):
* A special `memzero` function which avoids being removed by dead code elimination,
* a `mlock` function to prevent memory landing in swap or core dumps
* functions to prevent any access, read-only access or read-write access for a memory location
* a malloc which adds guard pages around allocated memory (see the sodium manual for details)

All these functions are used to prevent sensitive key material from leaking by some kind of buffer overflows or out of bound reads. Such memory 'leakage' is a much more common problem than situations where an attacker gains complete access to a server and could read keys from files. With certain protocols (e.g. noiseprotocol.org) you only need the private key in memory for a short time to derive a ephemeral key. After this you want your private key cleared from memory completely. All these functions also require that you know exactly where you keep your secret data in memory.

In libsodium you [explicitly manage the memory for hash state](https://download.libsodium.org/doc/hashing/generic_hashing.html) and hash result values so it's possible to use these secure memory functions.

So what's the problem with std.digest? The hash state is a struct, so it's pass by value. This can be fixed by a simple by-reference wrapper. But the `finish` methods can cause problems:

```d
ubyte[32] finish()
{
    ubyte[32] result = state;
    result[1] &= 42;
    return result;
}

void bar()
{
    auto secretKey = finish();
    // do something
    secureClearMemory(secretKey);
}
```

Without certain compiler optimization you might actually end up with two copies of the secret data: `result` might be in a different location on stack than `secretKey`. So the `secureClear` function will only clear one copy but the secret data will still be in stack memory. And depending on your function call patters and function stack sizes it might stay there for quite some time waiting for attackers scanning the applications memory....

Another problem is that you can't force the result value to be only in secure memory and never on the stack.

Now compiler optimizations, NRVO and the fact that large return values are usually passed by hidden reference (depends on the system ABI) usually help to avoid such copies. But I wouldn't want to rely on such optimizations for crypto code (Always remember that in certain countries in the worst case unsafe cryptography could be life-threatening).

Some related links:
https://www.securecoding.cert.org/confluence/display/cplusplus/MSC06-CPP.+Be+aware+of+compiler+optimization+when+dealing+with+sensitive+data
http://stackoverflow.com/questions/10683941/clearing-memory-securely-and-reallocations
http://www.cs.utsa.edu/~shxu/dsn07.pdf
http://stackoverflow.com/questions/1263350/cryptography-best-practices-for-keys-in-memory

And more related to the original pull request: A MSVC optimization which rendered a safe constant-time algorithm unsafe:
https://research.kudelskisecurity.com/2017/01/16/when-constant-time-source-may-not-save-you/

> Won't this also be an issue in any scripting language?

Probably, depends a lot on the language implementation.

**Comment by jpf91** [comment]
> Now, assume unpredictableSeed is cryptographically secure. Would that make this algorithm cryptographically secure?

Here's a link why you have to be especially careful when dealing with side-channel sensitive crypto code: http://cr.yp.to/talks/2014.10.18/slides-djb-20141018-a4.pdf (start at page 17)

Especially
> Try to scare implementors away from constant-time software. e.g. “It will be too slow.” “It’s too hard to write.”
[...]
> Fund variable-time software, maybe with “countermeasures” that make the timings difficult for researchers to analyze but that are still breakable with our computer resources.

This is a slippery slope. Better not go there ;-)

**Comment by quickfur** [comment]
@andralex: Relying on `Xorshift32`, which is *not* a CSPRNG, already makes this algorithm questionable, nevermind the fact that `unpredictableSeed` is not cryptographically secure (and can never be because its range of values is far too small).  Furthermore, the fact that failures are measurably faster than successes means that this algorithm will still fail under a similar side-channel attack.  You may have thwarted this specific instance of it (guess the secret byte-by-byte consecutively) but that does not mean the attacker cannot use other adaptations of the same idea (exploiting the performance difference between success and failure in comparison) to ferret out your secret.

It's usually a bad idea to try to outsmart recommendations from cryptography experts, because chances are they have *already* thought of what you had in mind and have dismissed it for various reasons.

**Comment by quickfur** [comment]
Also, the way `unpredictableSeed` is used here is exactly what was warned against by the recent article linked in the forums.

Imagine if somebody wrote an Apache module that serves requests by spawning a new thread per connection (a pretty common model). Suppose the first request that comes in per new connection is a login request -- a rather likely occurrence. That means the first time this comparison function is called is likely also the *only* time it will get called in that thread. This effectively means that every thread (login request) is calling `unpredictableSeed` once, and then extracting only a single value from the PRNG seeded by it.  Across the span of requests, then, you're effectively using `unpredictableSeed` as your RNG rather than the actual PRNG algorithm.  This is the textbook case of overseeding your PRNG to the point the PRNG itself becomes irrelevant (you could even use a CSPRNG here and the repeated seeding would effectively negate its security).

Seeding should never be done in a low-level function like this; it needs to be done globally and in a controlled way that does not nullify the PRNG it is seeding.

**Comment by andralex** [comment]
Fair enough, thanks all!

**Pull request merged**